### ISSUE TITLE
More sophisticated adblocking

### ISSUE THREAD

**Issue by The-Compiler** [qutebrowser/qutebrowser#29]
~~There should be an adblocker.~~

**edit: A host-based adblocker which is sufficient for most cases is implemented since a long time already, this issue is still open because it could be improved by implementing ABP-like filterlist support.**

The eric IDE helpviewer does have one (in Python), which might be a good inspiration

**Comment by 1-61803** [comment]
Is there any ad-blocking implementation as of now, internal or via plugin?

**Comment by parchd-1** [comment]
Yes, there is host-file adblocking, as you can see above.
The plugin system isn't done yet, so other adblock systems aren't either. You can also see that in the issue linked to above ;).

**Comment by 1-61803** [comment]
@parchd-1 I skimmed. I see the hosts file implementation from 2014 and #30 has adblock stroke through. What is the actual state of development of plugins that would allow better ad blocking but also plugins like ublock and greasemonkey?

**Comment by The-Compiler** [comment]
> What is the actual state of development of plugins

There isn't anything other than the thoughts listed in #30 so far.

> that would allow better ad blocking but also plugins like ublock

What from ublock are you missing, other than having adblockplus-like block lists?

> and greasemonkey

Something similar to it should be possible with per-domain settings (#27) which is the current main focus - in around a week I'll launch qutebrowser's second crowdfunding, to work on this (and maybe start some work on plugins if there's time left) during my summer holidays. Subscribe to that issue if you want to get notified once it's up!

**Comment by 1-61803** [comment]
> What from ublock are you missing, other than having adblockplus-like block lists?

I just have switched to ublock as ABP caused high CPU usage. Having the same lists makes rendering sites equally across hosts. Just a first impression after switching blocker.

I still have to make it work in my main older machine on 10.8. I encountered the same problem pointed out [here](https://github.com/qutebrowser/qutebrowser/issues/1357#issuecomment-213255634). It should compile from source, it's python and qt, right?

> Something similar to it should be possible with per-domain settings (#27) which is the current main focus - in around a week I'll launch qutebrowser's second crowdfunding

Subscribed.  What's the best way to get regular (weekly but not daily) news? **If** I compile it in my 10.8 I'll be certainly donating to your project.

**Comment by The-Compiler** [comment]
> I still have to make it work in my main older machine on 10.8. I encountered the same problem pointed out here. It should compile from source, it's python and qt, right?

There's nothing in qutebrowser itself you'll need to compile, as it's Python. You'll need a working Python with PyQt though. If you can get a Python 3.5 or 3.6 installed (maybe via Homebrew or [pyenv](https://github.com/pyenv/pyenv)), you should be able to do something like `pip install PyQt5 qutebrowser` (optionally in a [virtualenv](https://virtualenv.pypa.io/)). If you need help, open a separate issue though, please.

> What's the best way to get regular (weekly but not daily) news?

There isn't anything like a regular newsletter. You can look at the [changelog](https://github.com/qutebrowser/qutebrowser/blob/master/CHANGELOG.asciidoc) and [github activity](https://github.com/qutebrowser/qutebrowser/pulse) manually though.

I'd also recommend subscribing to the [qutebrowser-announce mailinglist](https://lists.schokokeks.org/mailman/listinfo.cgi/qutebrowser-announce) where important announcements like new releases are posted.

**Comment by asmbits** [comment]
Is there any progress? It is the only reason I haven't switched to qutebrowser yet. I cannot imagine seeing ads while browsing google, youtube or any other site.

**Comment by The-Compiler** [comment]
@asmbits I don't think anyone is working on this currently, as the current adblocker is working well enough overall and there are lots and lots of other more important issues :wink:

**Comment by sagb** [comment]
On the sites where ads come not from standard CDNs the current adblock is useless. This major bug is very annoying.

**Comment by The-Compiler** [comment]
You seem to have an... interesting understanding of what "major bug" means. I'd be glad if someone picked this up, but it won't be me in the near future, I have lots of other stuff on my plate. Are you volunteering? :wink:

**Comment by sagb** [comment]
Sorry, it's beyond my skills and available time.

**Comment by CosmoJG** [comment]
@The-Compiler Would it be alright if myself or someone else tried directly integrating some of uBlock's code? Or would you rather have something built from scratch?

I think direct integration would do wonders for preventing the classic open-source effort-reduplication problem.

**Comment by The-Compiler** [comment]
@CosmoJG I don't see how that'd be possible (as no WebExtension API exists), so the point is moot :wink:

**Comment by CosmoJG** [comment]
@The-Compiler Ahh, admittedly, I haven't put much thought into this, but I wasn't talking about including it as an extension. I meant I'd try and directly integrate the parts of uBlock Origin that make it different from Qutebrowser's built-in adblock. I'm operating on the assumption that there are code bits within Qutebrowser that perform similar functions to whatever WebExtension API calls uBlock makes.

Now that I think about this further, such a "compatibility layer" would probably come with loads of performance problems, and I'd be better off contributing to #30 so that Qutebrowser could support other plugins as well.

Maybe when I have less work later in the week, I'll get off my ass (and stop talking out of it) and dig through Qutebrowser's core adblock to see what kinds of original improvements can be made. Granted, I've never written code for adblocking before and only have a basic understanding of how most of it works, but that doesn't mean I can't try!

**Comment by jgkamat** [comment]
I too have little idea how adblocking (and js in general) work, but I think a good starting point for this is to see how [greasemonkey userscripts](https://github.com/qutebrowser/qutebrowser/blob/7eaad59be36e9f7a32c1806cf1bbdd3bbbbb8188/qutebrowser/browser/webengine/webenginetab.py#L722-L755) are injected into the page (and stylesheets as well). Once you get a script in the page, it becomes a js issue, which I have almost zero experience with :P 

I don't think the current host-blocking adblock code is a good place to start (because this would probably need a completely different approach), but [I think it's mostly located here](https://github.com/qutebrowser/qutebrowser/blob/master/qutebrowser/browser/adblock.py).

**Comment by The-Compiler** [comment]
You won't be able to block network requests with JS injected in the page - there's stuff in the WebExtension API to do that which uBlock uses, but doing the same in qutebrowser won't be possible. In qutebrowser, JS injected into the page can only do whatever the page itself can do, and it has the same JS API a webpage can use (and not more).

Adblocking for qutebrowser really has to be implemented in Python - JavaScript won't work without [having relevant APIs in QtWebEngine](https://bugreports.qt.io/browse/QTBUG-61676) which probably won't be happening in the near future.

So, all you'll be able to take from uBlock is some general ideas and concepts (i.e. the "big picture"), and the filter list syntax (which is mostly coming from AdBlockPlus) - but not any actual code.

**Comment by rr-** [comment]
The built-in filters seem to be less effective with each month, despite `:adblock-update` :(

**Comment by Yodzorah** [comment]
1. What about requests blocking via our two backends? Do they support domain patterns? And if not, what about `dnsmasq`?

2. uBlock, AdBlock, etc. suck. Sorry, but it's true. They work on similar rulesets for request blocking (with patterns, though), css and js and they are centered on just removing ads. They have full control over files, scripts, etc., they have full functionality for full control over the mentioned stuff, but instead of providing the control to user they just block stuff according to rulesets. It's just a pointless waste of resources. Look at ContentBlockHelper and uMatrix - they are much better. In qutebrowser we have JS, CSS and hosts blocklists. Also, we probably have most of the features of ContentBlockHelper/uMatrix in QtWebEngine, we just need to wait for per domain settings. I think, if there is any chance we can use more simple and powerful solutions instead of emulating familiar to many users but grotesque and ailing adblokers, we should think trice before choosing the second option.

3. Look at the ContentBlockHelper UI. I mean, not the design, it's awful, but the tree view of every single file for every single website. What about something like that in qutebrowser? For example, we have ':set' mode, can we have the same for css, js, hosts, cookies, etc. settings in "per domain" mode? Not just with a sublist of settings but with the website tree?

**Comment by The-Compiler** [comment]
@Yodzorah 

1) See #4188, though real adblock lists would also be patterns. Not sure what you're asking about dnsmasq.

2) You're free to have your opinion, but not everyone wants to spend time configuring rules on every website they visit. I'd like to have both in the long run.

3) That sounds quite similar to uMatrix, see #28 (and also #2626)

**Comment by Yodzorah** [comment]
@The-Compiler,

2. My option doesn't require users to configure websites all the time in order to get rid from ads. It can do both with the functionality of usual adblockers, while the usual adblockers with functionality of usual adblockers can only block ads. I just don't think it's a good idea to have such highly specialized tool that, in fact, can do more really useful stuff with the same 'effort'. If one will add +10% of functionality to any popular adbloker, he will have far more powerfull and flexible tool and far more control over web. And if one will use usual 'simple' adblockers and suddenly understand he wants to block something else, he will ask for another highly specialized tool instead of something more flexible for both tasks. It's like installing separate extensions changing CSS for some websites instead of installing Stylish in Chrome. 

Sorry for asking, but if qutebrowser is so centered on providing user full control, why prefer highly specialized solutions over flexible? I mean, not everyone even knows about user agent header, why such setting even exists in qutebrowser then?

**Comment by The-Compiler** [comment]
I'm not sure what you're trying to achieve with this discussion - like I said above I'd like to have both, as both solutions have their benefits and drawbacks. This is not the right place to discuss whether adblockers are a good thing or not.

**Comment by Yodzorah** [comment]
@The-Compiler, I was trying to convince you to not try usual adblockers until you will be sure there is no flexible solution) Maybe, it was not wise, but I should have try at least - I'm using qutebrowser because of its flexibility, so if this won't be the case, I will lose even a half-comfortable browser and will be forced to migrate on Vivaldi, because with the same low level of flexibility it at least has a big pack of features.

**Comment by toofar** [comment]
There is no plan for qutebrowser to support the usual adblockers. See https://github.com/qutebrowser/qutebrowser/issues/2626 for a discussion about a flexible request blocking framework and https://github.com/qutebrowser/qutebrowser/issues/28 for a prettier UI over that.

**Comment by The-Compiler** [comment]
Note I've already linked those [above](https://github.com/qutebrowser/qutebrowser/issues/29#issuecomment-419400209), but whatever...

**Comment by mschilli87** [comment]
Is there any way we will ever see this given [Chromium might loose a lot of its content blocking support](https://www.theregister.co.uk/2019/01/22/google_chrome_browser_ad_content_block_change/)?
It's a pity these things happen just when [we might loose support for any alternative engine](https://github.com/qutebrowser/qutebrowser/issues/4039).

**Comment by jgkamat** [comment]
> Is there any way we will ever see this given [Chromium might loose a lot of its content blocking support](https://www.theregister.co.uk/2019/01/22/google_chrome_browser_ad_content_block_change/)?
> It's a pity these things happen just when [we might loose support for any
> alternative engine](https://github.com/qutebrowser/qutebrowser/issues/4039).

FWIW, I don't think that can affect anything in qutebrowser, given that
qutebrowser can't use the webextension api at all. Any blocking will occur
through the requestInterceptor class instead, so as long as qt continues
supporting that, we shouldn't have any problem:

https://doc.qt.io/qt-5/qwebengineurlrequestinterceptor.html

A bigger issue is that the api to block requests in qtwebkit seems to not
expose any information about the request, making it impossible to do blocking
based on request type (uMatrix).

The non-viral first party source for that news is here:

https://docs.google.com/document/d/1nPu6Wy4LWR66EFLeYInl3NzzhHzc-qnk4w4PX-0XMw8/edit#

**Comment by jgkamat** [comment]
I spent the last couple of days knocking out a prototype pure-python adblocker
based off the high level algorithm in abp/ubo (with a few extra tricks) using
the prototype interceptor api. It won't support element hiding (that should be
a completely separate project) and also won't understand any context until
#4525. Right now, it seems to be in-between ubo and abp in terms of latency,
but I think it can get a lot better than that, as there are a still a few
glaring issues with my implementation.

However, I don't really see much of a difference (I couldn't find many ads to
test on), I guess it shows how great steven black's host list is. I personally
don't think it's worth the cpu/memory/latency cost anymore, considering ubo
claims to be over an order of magnitude slower than the current blocker, and
most video ads can be bypassed (using a better interface) with mpv.

Here's a demo: https://www.youtube.com/watch?v=Bd29bqfuCSc
Source: https://gitlab.com/jgkamat/jblock

Also (probably more importantly), the tokenizer and bucketing logic from there
(after they're improved) can be used as a trivial way to solve #4409, as our
url patterns are a much easier version of abp style blocklists.

**Comment by CosmoJG** [comment]
@jgkamat Jblock looks awesome! How far away are you from implementing this in Qutebrowser? How does it fare against webmacs's implementation of brave/ad-block?

**Comment by jgkamat** [comment]
I happened to come across Snuupy who has done a lot of work with element
hiding in brave (this is a synopsis of what I learned)

1. Here's an example of syntax we can use (I think) to let the browser decide
when to hide elements (along with the strategy ublock uses to filter them
down):

https://github.com/uBlock-LLC/uBlock/issues/161

I *think* that means we won't need per-domain css (the current implementation
of that is really hacky), but I could be wrong. Without per-domain css, we
wouldn't be able to filter them down as easily like ublock does (it would need
to be done in js), so we would need to add all rules to every page, which
could be slow.

I've heard that abp adds all the rules to the shadowdom every time.

2. Here's the commit adding basic element hiding via css to brave:

https://github.com/brave/brave-extension/commit/b914102c37fa8deba7ed5bb8b08185c536e82d48

3. CSS element hiding is not sufficient for all cases, as some sites will add
!important to the html tag, and html !important trumps css !important. To get
around that, some adblockers collapse/hide the element in the dom. I don't
think we should worry about that now unless we wanted to be very pedantic
though, css would be easier imo.

4. This branch has a wip version of the non-css method:
https://github.com/Snuupy/brave-extension/tree/mutationObserver

**Comment by The-Compiler** [comment]
https://whotracks.me/blog/adblockers_performance_study.html has some interesting performance analysis of different adblocker engines.

All those engines being available implemented in JavaScript gave me a somewhat crazy idea: Could we use a [QJSEngine](https://doc.qt.io/qt-5/qjsengine.html) to run a JavaScript adblocker and use its decisions (allow/block) from Python?

**Comment by The-Compiler** [comment]
From @jgkamat in IRC: https://testpages.adblockplus.org/ is quite useful to test an implementation.

**Comment by jgkamat** [comment]
I've implemented a simple script to handle element blocking through
greasemonkey (with GM_addStyle). Right now, its rather trivial as it adds all
global rules to pages which may be less efficient (definitely in memory, maybe
in cpu) than filtering them in js beforehand. The main improvement is that it
removes the empty space left behind by most network blocks, for example, the
youtube and reddit banner ads (although whether that's worth the cost is
questionable).

Right now it already supports global element hiding rules and should properly
handle site specific overrides and exceptions. I'm doing this with a second
overriding rule, rather than not applying the rule at all for sim (for
simplicity), which may be a small edge case.

At the very least, it passes:

- https://testpages.adblockplus.org/en/filters/element-hiding
- https://testpages.adblockplus.org/en/exceptions/element-hiding

And it wont (and probably never will) pass the extended selectors test:

- https://testpages.adblockplus.org/en/filters/element-hiding-emulation

https://gitlab.com/jgkamat/jhide

**Comment by zwykl3** [comment]
How using my own adblocker in qutebrowser?
I need oly small file with adress. Nothing more. But I need delete domain
*.gemius.pl for example

**Comment by The-Compiler** [comment]
@zwykl3 There isn't any extension API yet, see #30. You can add your own hosts to e.g. `~/.config/qutebrowser/blocked-hosts` but it doesn't support wildcards.

**Comment by zwykl3** [comment]
1. I'm noth have directory ~/.config/qutebrowser/
(on mac)
2. is possible to add wilcards?

In my opinion this will be good if it work

**Comment by The-Compiler** [comment]
> I'm noth have directory ~/.config/qutebrowser/

The config directory is listed in `:version`

> is possible to add wilcards?

Not currently.

**Comment by The-Compiler** [comment]
Looks like brave is working on a Rust-based adblocking library:

- [Blogpost](https://brave.com/improved-ad-blocker-performance/)
- [Repository](https://github.com/brave/adblock-rust)

**Comment by The-Compiler** [comment]
Patch from #4543 using the (old) Brave C adblock:

```diff
From 3ed1400e157a22eedd4ca128151586eee61aa5e1 Mon Sep 17 00:00:00 2001
From: Stephane ALBERT <stephane.albert@amadeus.com>
Date: Wed, 23 Jan 2019 13:05:30 +0100
Subject: [PATCH] WIP: Experimental support for brave adblocker

---
 .gitmodules                       |   9 ++
 c/adblock.c                       | 237 ++++++++++++++++++++++++++++++
 qutebrowser/components/brave.py   | 169 +++++++++++++++++++++
 qutebrowser/config/configdata.yml |  17 +++
 setup.py                          |  33 ++++-
 vendor/ad-block                   |   1 +
 vendor/bloom-filter-cpp           |   1 +
 vendor/hashset-cpp                |   1 +
 8 files changed, 467 insertions(+), 1 deletion(-)
 create mode 100644 .gitmodules
 create mode 100644 c/adblock.c
 create mode 100644 qutebrowser/components/brave.py
 create mode 160000 vendor/ad-block
 create mode 160000 vendor/bloom-filter-cpp
 create mode 160000 vendor/hashset-cpp

diff --git a/.gitmodules b/.gitmodules
new file mode 100644
index 0000000000..d52b9d2f0c
--- /dev/null
+++ b/.gitmodules
@@ -0,0 +1,9 @@
+[submodule "vendor/hashset-cpp"]
+	path = vendor/hashset-cpp
+	url = https://github.com/bbondy/hashset-cpp
+[submodule "vendor/bloom-filter-cpp"]
+	path = vendor/bloom-filter-cpp
+	url = https://github.com/bbondy/bloom-filter-cpp
+[submodule "vendor/ad-block"]
+	path = vendor/ad-block
+	url = https://github.com/brave/ad-block
diff --git a/c/adblock.c b/c/adblock.c
new file mode 100644
index 0000000000..5fedac7c51
--- /dev/null
+++ b/c/adblock.c
@@ -0,0 +1,237 @@
+// This file is part of webmacs.
+//
+// webmacs is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// webmacs is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with webmacs.  If not, see <http://www.gnu.org/licenses/>.
+
+#include <Python.h>
+#include "structmember.h"
+
+#include <iostream>
+#include <fstream>
+
+#include "ad_block_client.h"
+
+using namespace std;
+
+typedef struct {
+  PyObject_HEAD
+
+  AdBlockClient * client;
+  char * data;
+} AdBlock;
+
+
+static void
+AdBlock_dealloc(AdBlock* self)
+{
+  delete self->client;
+  if (self->data) delete[] self->data;
+  Py_TYPE(self)->tp_free((PyObject*)self);
+}
+
+static PyObject *
+AdBlock_new(PyTypeObject *type, PyObject *args, PyObject *kwds)
+{
+  AdBlock *self;
+
+  self = (AdBlock *)type->tp_alloc(type, 0);
+
+  return (PyObject *)self;
+}
+
+static int
+AdBlock_init(AdBlock *self, PyObject *args, PyObject *kwds)
+{
+  self->client = new AdBlockClient;
+  self->data = NULL;
+  return 0;
+}
+
+static PyObject *
+AdBlock_parse(AdBlock* self, PyObject *args)
+{
+  const char *data;
+
+  if (!PyArg_ParseTuple(args, "s", &data))
+    return NULL;
+
+  Py_BEGIN_ALLOW_THREADS
+  self->client->parse(data);
+  Py_END_ALLOW_THREADS
+
+  Py_RETURN_NONE;
+}
+
+static PyObject *
+AdBlock_matches(AdBlock* self, PyObject *args)
+{
+  const char *url, *domain;
+  bool result;
+
+  if (!PyArg_ParseTuple(args, "ss", &url, &domain))
+    return NULL;
+
+  /* I suspect that allowing threads here does create deadlocks,
+   but anyway I am not sure it would be useful to allow them. */
+  /* Py_BEGIN_ALLOW_THREADS */
+  result = self->client->matches(url, FONoFilterOption, domain);
+  /* Py_END_ALLOW_THREADS */
+
+  if (result) {
+    Py_RETURN_TRUE;
+  } else {
+    Py_RETURN_FALSE;
+  }
+}
+
+static PyObject *
+AdBlock_save(AdBlock* self, PyObject *args)
+{
+  const char *path;
+
+  if (!PyArg_ParseTuple(args, "s", &path))
+    return NULL;
+
+  int size;
+  ofstream outFile(path, ios::out | ios::binary);
+  if (!outFile) {
+    Py_RETURN_FALSE;
+  }
+
+  Py_BEGIN_ALLOW_THREADS
+  char * buffer = self->client->serialize(&size);
+  outFile.write(buffer, size);
+  outFile.close();
+  Py_END_ALLOW_THREADS
+
+  Py_RETURN_TRUE;
+}
+
+static PyObject *
+AdBlock_load(AdBlock* self, PyObject *args)
+{
+  const char *path;
+  bool result = false;
+
+  if (!PyArg_ParseTuple(args, "s", &path))
+    return NULL;
+
+  ifstream file(path, ios::binary | ios::ate);
+  if (!file) {
+    Py_RETURN_FALSE;
+  }
+
+  Py_BEGIN_ALLOW_THREADS
+  streamsize size = file.tellg();
+  file.seekg(0, ios::beg);
+
+  if (self->data) {delete[] self->data;}
+  self->data = new char[size];
+  if (file.read(self->data, size)) {
+    self->client->deserialize(self->data);
+    result = true;
+  }
+  Py_END_ALLOW_THREADS
+
+  if (result) {
+    Py_RETURN_TRUE;
+  } else {
+    Py_RETURN_FALSE;
+  }
+}
+
+
+static PyMethodDef AdBlock_methods[] = {
+  {"parse", (PyCFunction)AdBlock_parse, METH_VARARGS,
+   "Parse adblock data string, like the content of an easylist."
+  },
+  {"matches", (PyCFunction)AdBlock_matches, METH_VARARGS,
+   "matches an url, returns True if it should be filtered."
+  },
+  {"save", (PyCFunction)AdBlock_save, METH_VARARGS,
+   "Save serialized data into a file."
+  },
+  {"load", (PyCFunction)AdBlock_load, METH_VARARGS,
+   "Load serialized data from a file."
+  },
+  {NULL}  /* Sentinel */
+};
+
+static PyTypeObject AdBlockType = {
+    PyVarObject_HEAD_INIT(NULL, 0)
+    "adblock.AdBlock",             /* tp_name */
+    sizeof(AdBlock),             /* tp_basicsize */
+    0,                         /* tp_itemsize */
+    (destructor)AdBlock_dealloc, /* tp_dealloc */
+    0,                         /* tp_print */
+    0,                         /* tp_getattr */
+    0,                         /* tp_setattr */
+    0,                         /* tp_reserved */
+    0,                         /* tp_repr */
+    0,                         /* tp_as_number */
+    0,                         /* tp_as_sequence */
+    0,                         /* tp_as_mapping */
+    0,                         /* tp_hash  */
+    0,                         /* tp_call */
+    0,                         /* tp_str */
+    0,                         /* tp_getattro */
+    0,                         /* tp_setattro */
+    0,                         /* tp_as_buffer */
+    Py_TPFLAGS_DEFAULT |
+        Py_TPFLAGS_BASETYPE,   /* tp_flags */
+    "Adblock objects",           /* tp_doc */
+    0,                         /* tp_traverse */
+    0,                         /* tp_clear */
+    0,                         /* tp_richcompare */
+    0,                         /* tp_weaklistoffset */
+    0,                         /* tp_iter */
+    0,                         /* tp_iternext */
+    AdBlock_methods,             /* tp_methods */
+    0,             /* tp_members */
+    0,                         /* tp_getset */
+    0,                         /* tp_base */
+    0,                         /* tp_dict */
+    0,                         /* tp_descr_get */
+    0,                         /* tp_descr_set */
+    0,                         /* tp_dictoffset */
+    (initproc)AdBlock_init,      /* tp_init */
+    0,                         /* tp_alloc */
+    AdBlock_new,                 /* tp_new */
+};
+
+
+static PyModuleDef adblockmodule = {
+  PyModuleDef_HEAD_INIT,
+  "adblock",
+  "Module to speed up ad filtering.",
+  -1,
+  NULL, NULL, NULL, NULL, NULL
+};
+
+PyMODINIT_FUNC
+PyInit__adblock(void)
+{
+  PyObject* m;
+
+  AdBlockType.tp_new = PyType_GenericNew;
+  if (PyType_Ready(&AdBlockType) < 0)
+    return NULL;
+
+  m = PyModule_Create(&adblockmodule);
+  if (m == NULL)
+    return NULL;
+
+  Py_INCREF(&AdBlockType);
+  PyModule_AddObject(m, "AdBlock", (PyObject *)&AdBlockType);
+  return m;
+}
diff --git a/qutebrowser/components/brave.py b/qutebrowser/components/brave.py
new file mode 100644
index 0000000000..1656ee8bf0
--- /dev/null
+++ b/qutebrowser/components/brave.py
@@ -0,0 +1,169 @@
+# vim: ft=python fileencoding=utf-8 sts=4 sw=4 et:
+
+# Copyright 2014-2018 Florian Bruhin (The Compiler) <mail@qutebrowser.org>
+#
+# This file is part of qutebrowser.
+#
+# qutebrowser is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# qutebrowser is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with qutebrowser.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Functions related to ad blocking."""
+
+import io
+import os.path
+import functools
+import posixpath
+import zipfile
+import logging
+import typing
+import pathlib
+
+from PyQt5.QtCore import QUrl
+
+from _adblock import AdBlock
+from qutebrowser.api import (cmdutils, hook, config, message, downloads,
+                             interceptor, apitypes)
+
+
+logger = logging.getLogger('misc')
+_brave_adblocker = typing.cast('BraveAdBlocker', None)
+
+
+class BraveAdBlocker:
+
+    """Manage blocked hosts based from /etc/hosts-like files.
+
+    Attributes:
+        _in_progress: The DownloadItems which are currently downloading.
+        _done_count: How many files have been read successfully.
+        _has_basedir: Whether a custom --basedir is set.
+    """
+
+    def __init__(self, *, data_dir: pathlib.Path, config_dir: pathlib.Path,
+                 has_basedir: bool = False) -> None:
+        self._has_basedir = has_basedir
+        self._in_progress = []  # type: typing.List[downloads.TempDownload]
+        self._done_count = 0
+        self._local_adblock_cache = str(data_dir / 'adblock-cache.dat')
+        self._adblock = AdBlock()
+
+    def _is_blocked(self, request_url: QUrl,
+                    first_party_url: QUrl = None) -> bool:
+        """Check whether the given request is blocked."""
+        if not config.get('content.brave_adblock.enabled', False):
+            return False
+        if first_party_url is not None and not first_party_url.isValid():
+            return False
+        url_s = request_url.toString()
+        fp_host = first_party_url.host()
+        if fp_host and self._adblock.matches(url_s, fp_host):
+            return True
+        return False
+
+    def filter_request(self, info: interceptor.Request) -> None:
+        """Block the given request if necessary."""
+        if self._is_blocked(request_url=info.request_url,
+                            first_party_url=info.first_party_url):
+            logger.info(
+                "Request to {} blocked by brave adblocker.".format(
+                    info.request_url.toString()))
+            info.block()
+
+    def read_cache(self) -> None:
+        """Read hosts from the existing blocked-hosts file."""
+
+        if not os.path.isfile(self._local_adblock_cache):
+            if (config.val.content.brave_adblock.lists and
+                    not self._has_basedir and
+                    config.val.content.brave_adblock.enabled):
+                message.info("Run :brave-adblock-update to get adblock lists.")
+        else:
+            logging.info("loading adblock cached data: %s", self._local_adblock_cache)
+            self._adblock.load(self._local_adblock_cache)
+
+    def adblock_update(self) -> None:
+        """Update the adblock block lists."""
+        self._done_count = 0
+        logging.info("Downloading new blocklists.")
+        for url in config.val.content.brave_adblock.lists:
+            logging.info("Adding: %s" % url.toString())
+            download = downloads.download_temp(url)
+            self._in_progress.append(download)
+            download.finished.connect(
+                functools.partial(self._on_download_finished, download))
+
+    def _on_lists_downloaded(self) -> None:
+        """Install adblock lists after files have been downloaded."""
+        message.info("brave adblock: Saving cache file.")
+        self._adblock.save(self._local_adblock_cache)
+
+    def update_files(self) -> None:
+        """Update files when the config changed."""
+        if not config.val.content.brave_adblock.lists:
+            try:
+                os.remove(self._local_adblock_cache)
+            except FileNotFoundError:
+                pass
+            except OSError as e:
+                logger.exception("Failed to delete adblock cache file: {}".format(e))
+
+    def _on_download_finished(self, download: downloads.TempDownload) -> None:
+        """Check if all downloads are finished and if so, trigger reading.
+
+        Arguments:
+            download: The finished download.
+        """
+        self._in_progress.remove(download)
+        if download.successful:
+            self._done_count += 1
+            try:
+                download.fileobj.seek(0)
+                text = io.TextIOWrapper(download.fileobj)
+                self._adblock.parse(text.read())
+                text.close()
+            finally:
+                download.fileobj.close()
+        if not self._in_progress:
+            try:
+                self._on_lists_downloaded()
+            except OSError:
+                logger.exception("Failed to write brave adblock cache!")
+
+
+@cmdutils.register()
+def brave_adblock_update() -> None:
+    """Update the adblock block lists.
+
+    This updates `~/.local/share/qutebrowser/blocked-hosts` with downloaded
+    host lists and re-reads `~/.config/qutebrowser/blocked-hosts`.
+    """
+    # FIXME: As soon as we can register instances again, we should move this
+    # back to the class.
+    _brave_adblocker.adblock_update()
+
+
+@hook.config_changed('content.brave_adblock.lists')
+def on_config_changed() -> None:
+    _brave_adblocker.update_files()
+
+
+@hook.init()
+def init(context: apitypes.InitContext) -> None:
+    """Initialize the brave adblocker."""
+    global _brave_adblocker
+    _brave_adblocker = BraveAdBlocker(
+        data_dir=context.data_dir,
+        config_dir=context.config_dir,
+        has_basedir=context.args.basedir is not None)
+    _brave_adblocker.read_cache()
+    interceptor.register(_brave_adblocker.filter_request)
diff --git a/qutebrowser/config/configdata.yml b/qutebrowser/config/configdata.yml
index b6d4736ae9..8d4c26295c 100644
--- a/qutebrowser/config/configdata.yml
+++ b/qutebrowser/config/configdata.yml
@@ -523,6 +523,23 @@ content.headers.user_agent:
 
     Note that the value read from JavaScript is always the global value.
 
+content.brave_adblock.enabled:
+  default: true
+  supports_pattern: true
+  type: Bool
+  desc: Enable Brave AdBlock.
+
+content.brave_adblock.lists:
+  default:
+    - "https://easylist.to/easylist/easylist.txt"
+    - "https://easylist.to/easylist/fanboy-annoyance.txt"
+  type:
+    name: List
+    valtype: Url
+    none_ok: true
+  desc: |
+    List of URLs of lists which contain adblock rules.
+
 content.host_blocking.enabled:
   default: true
   supports_pattern: true
diff --git a/setup.py b/setup.py
index 226b0a73ee..c5d4da43e8 100755
--- a/setup.py
+++ b/setup.py
@@ -32,11 +32,41 @@
 
 
 try:
-    BASEDIR = os.path.dirname(os.path.realpath(__file__))
+    BASEDIR = os.path.dirname(os.path.relpath(__file__))
 except NameError:
     BASEDIR = None
 
 
+bloom_dir = os.path.join(BASEDIR, "vendor", "bloom-filter-cpp")
+hashset_dir = os.path.join(BASEDIR, "vendor", "hashset-cpp")
+adblock_dir = os.path.join(BASEDIR, "vendor", "ad-block")
+
+
+if "CC" not in os.environ:
+    # force g++, not sure why but else gcc is used and the code does not
+    # compile...
+    os.environ["CC"] = "g++"
+
+adblocker = setuptools.Extension(
+    '_adblock',
+    define_macros=[],
+    language="c++",
+    include_dirs=[bloom_dir, hashset_dir, adblock_dir],
+    # not sure if that help for speed. Careful it strip the debug symbols
+    extra_compile_args=["-g0", "-std=c++11"],
+    sources=[
+        os.path.join(bloom_dir, "BloomFilter.cpp"),
+        os.path.join(bloom_dir, "hashFn.cpp"),
+        os.path.join(hashset_dir, "hash_set.cc"),
+        os.path.join(adblock_dir, "ad_block_client.cc"),
+        os.path.join(adblock_dir, "filter.cc"),
+        os.path.join(adblock_dir, "cosmetic_filter.cc"),
+        os.path.join(adblock_dir, "no_fingerprint_domain.cc"),
+        os.path.join(adblock_dir, "protocol.cc"),
+        os.path.join(BASEDIR, "c", "adblock.c"),
+    ])
+
+
 def read_file(name):
     """Get the string contained in the file named name."""
     with common.open_file(name, 'r', encoding='utf-8') as f:
@@ -102,6 +132,7 @@ def _get_constant(name):
             'Topic :: Internet :: WWW/HTTP :: Browsers',
         ],
         keywords='pyqt browser web qt webkit qtwebkit qtwebengine',
+        ext_modules=[adblocker],
     )
 finally:
     if BASEDIR is not None:
diff --git a/vendor/ad-block b/vendor/ad-block
new file mode 160000
index 0000000000..38be4ceb21
--- /dev/null
+++ b/vendor/ad-block
@@ -0,0 +1 @@
+Subproject commit 38be4ceb216a8093a7a2b1f17ea83da371ede3f7
diff --git a/vendor/bloom-filter-cpp b/vendor/bloom-filter-cpp
new file mode 160000
index 0000000000..635780bbed
--- /dev/null
+++ b/vendor/bloom-filter-cpp
@@ -0,0 +1 @@
+Subproject commit 635780bbedff137a6a83ec23871944e22069de5b
diff --git a/vendor/hashset-cpp b/vendor/hashset-cpp
new file mode 160000
index 0000000000..f86b0a5752
--- /dev/null
+++ b/vendor/hashset-cpp
@@ -0,0 +1 @@
+Subproject commit f86b0a5752545274e32c0dbb654c3592cc131c8a
```

Like mentioned there, I'd prefer the C/Rust <-> Python bridge to be a separate module outside of qutebrowser.

**Comment by treeshateorcs** [comment]
so if I want to give the rust adblocker a shot, what should i do?

**Comment by The-Compiler** [comment]
@treeshateorcs You'd need to write Python bindings for it (e.g. using [PyO3](https://github.com/PyO3/PyO3)), then use those from qutebrowser similar to [the current adblocking module](https://github.com/qutebrowser/qutebrowser/blob/master/qutebrowser/components/adblock.py)).

**Comment by ArniDagur** [comment]
I'm interested in creating Python bindings for Brave's Rust adblocker and integrating them into qutebrowser, as discussed above. @The-Compiler Has anything changed since this was discussed, that would be a reason for me not to proceed?

**Comment by The-Compiler** [comment]
@ArniDagur Sounds great! Nothing changed as far as I know. You might also want to set up [cibuildwheel](https://github.com/joerick/cibuildwheel) to build Python wheels (binary packages) of the Rust adblocker and Python bindings.

**Comment by ArniDagur** [comment]
See #5317

**Comment by Sevensidedmarble** [comment]
Just wanted to say I'm so pumped for #5317 to be merged. Host adblocking never felt quite 'right' to me and kept me going back to chromium.

**Comment by gordon-quad** [comment]
I have a question about #5317.
It mentions that it supports ABP syntax but it looks like adblock-rust repo mentions only URL matching against easylist-like filters. So my use-case is a little bit different, I use uBlock-like extension not only to block ads but also quite frequently use "block element" item from popup menu to block unwanted UI elements that might originate from website itself (e.g. modal "register right now" popups, cookie banners etc.). So will I be able to do so with #5317 merged or shall I create a separate issue/feature request for that?

**Comment by ArniDagur** [comment]
>  I use uBlock-like extension not only to block ads but also quite frequently use "block element" item from popup menu to block unwanted UI elements that might originate from website itself (e.g. modal "register right now" popups, cookie banners etc.). So will I be able to do so with #5317 merged or shall I create a separate issue/feature request for that?

That is called cosmetic blocking, and is supported by the underlying adblocking library. See https://github.com/brave/adblock-rust/blob/master/src/cosmetic_filter_cache.rs for example.

In the comments of #5317 we discussed implementing cosmetic filtering in qutebrowser (see https://github.com/qutebrowser/qutebrowser/pull/5317#issuecomment-612554553), The conclusion was that there would need to be more work done on the underlying infrastructure to support cosmetic blocking. The aforementioned pull request therefore only tackles network filtering.

**Comment by The-Compiler** [comment]
FWIW there's also the `delete` target for `:hint`, though that isn't persisted.

**Comment by jgkamat** [comment]
Árni Dagur writes:

> In the comments of #5317 we discussed implementing cosmetic filtering in 
> qutebrowser (see 
> https://github.com/qutebrowser/qutebrowser/pull/5317#issuecomment-612554553), 
> The conclusion was that there would need to be more work done on the 
> underlying infrastructure to support cosmetic blocking. The aforementioned 
> pull request therefore only tackles network filtering.

Note that it's been possible for year(s) to implement cosmetic filtering 
through greasemonkey - for anyone who wants it there shouldn't be any blockers 
other than qt bugs.

https://github.com/qutebrowser/qutebrowser/issues/29#issuecomment-481107331

**Comment by gordon-quad** [comment]
Thanks, I will take a look at it, that sounds like maybe a suboptimal from performance standpoint but a solution nonetheless.

**Comment by jgkamat** [comment]
gordon-quad writes:

> Thanks, I will take a look at it, that sounds like maybe a suboptimal from 
> performance standpoint but a solution nonetheless.

suboptimal in what way exactly? I don't think qb is doing anything 
intentionally to slow down scripts (to my knowledge at least) - and if it was 
that should be fixed. I don't really see how any alternative available would 
launch scripts any faster unless the greasemonkey wrappers were significant.

**Comment by crocket** [comment]
The userscript is slow.

**Comment by jgkamat** [comment]
crocket writes:

> The userscript is slow.

In what way, and how can it be fixed in qb? I don't think the greasemonkey 
wrapper would run much js much slower than an extension, although I could be 
wrong.

**Comment by crocket** [comment]
You mentioned

> Right now, we load all global styles on page load, which can add significant overhead (~10ms in my testing). Doing this differently may be more efficient, but would make it harder to measure the overhead.

**Comment by jgkamat** [comment]
crocket writes:

> You mentioned
>
>> Right now, we load all global styles on page load, which can add 
>> significant overhead (~10ms in my testing). Doing this differently may be 
>> more efficient, but would make it harder to measure the overhead.

Yes, that's actually the way ABP does (or used?) to do cosmetic filtering. I 
found the overhead negligable compared to other parts of qutebrowser unrelated 
to greasemonkey (which were orders of magnitude slower), so I didn't see a 
need to optimize it when there were much bigger saves elsewhere. It's not a 
blocker to cosmetic filtering - it should be easy to do better if you would 
like to!

**Comment by crocket** [comment]
I think jhide is a bit inconvenient because I can just save cosmetic filter rules in uBlock Origin.

**Comment by iio7** [comment]
Is there a reason why the ad blocker isn't working on YouTube even though it is blocked in the Steve Black hosts list?

**Comment by iio7** [comment]
Just found the answer here, so never mind: https://old.reddit.com/r/qutebrowser/comments/hv085s/why_does_the_adblock_not_work_also_how_do_i_turn/

**Comment by The-Compiler** [comment]
It's taken a while, but @ArniDagur's ad blocking integration based on Brave's Rust library has been merged! :tada: :sparkles:

**Comment by edio** [comment]
@The-Compiler , do you plan to release a new version some time soon?

**Comment by cole-h** [comment]
@edio See https://github.com/qutebrowser/qutebrowser/pull/5317#issuecomment-749732193, specifically:

> There's some follow-up work I want to do here (e.g. announcing this via mailinglist and reddit and such), but not today

**Comment by The-Compiler** [comment]
This will be released as part of v2.0.0, which will still take a while - there's [still various work to do](https://github.com/qutebrowser/qutebrowser/issues?q=is%3Aopen+is%3Aissue+milestone%3Av2.0.0) (also see #5749), and I also want to give this some time to settle (and perhaps catch regressions).
